### DISCUSSION TITLE
Can not import ESM bundle for browser with NextJS 14

### DISCUSSION THREAD

**Discussion by tembra** [vladmandic/human#420] [Q&A]
**Issue Description**

When importing Human on a NextJS 14 project, the framework tries to import the node bundle instead of ESM for browsers, despite adding `'use client'` on top of the page file.

Since node bundle requires a tfjs backend, the error below comes in:
```
./node_modules/@vladmandic/human/dist/human.node.js:7:1678
Module not found: Can't resolve '@tensorflow/tfjs-node'
```

I know that this is not an issue within the Human package at all and it's about NextJS SSR. However I guess you may help me to achieve the desired solution.

I already tried to load Human within a separated dynamic loaded component with `dynamic(() => import('@/components/Human'), { ssr: false })`, but it also did not work.

I also tried to import it inside a `useEffect` of the `Human` component created as you did in your class constructor in NextJS demo, but it also did not work.

I also tried to import Human explicitly from `@vladmandic/human/dist/human.esm` but it did not found the path.

Everything works fine if I use Vite instead of NextJS to create my project.

**Steps to Reproduce**

1. Create a new project with `npx create-next-app app-name` and choose all default configs.
2. Run `npm i`
3. Change `app/page.tsx` to add `'use client'` on top the file and add import to Human `import { Human } from '@vladmandic/human`.
4. Run `npm run dev`
5. Access the URL on your browser

**Expected Behavior**

NextJS identify that I'm using a client component on that page and load ESM bundle of Human package instead of node bundle.

**Environment**

- Human library version? 3.2.0
- Built-in demo or custom code? Custom code. Just create a project and import Human.
- Type of module used (e.g. `js`, `esm`, `esm-nobundle`)? esm
- TensorFlow/JS version (if not using bundled module)?
- Browser or NodeJS and version (e.g. *NodeJS 14.15* or *Chrome 89*)? Arc 1.28.0
- OS and Hardware platform (e.g. *Windows 10*, *Ubuntu Linux on x64*, *Android 10*)? macOS 14.2.1 on M2 Pro (arm64)
- Packager (if any) (e.g, *webpack*, *rollup*, *parcel*, *esbuild*, etc.)? 
- Framework (if any) (e.g. *React*, *NextJS*, etc.)? NextJS

**Diagnostics**

- Check out any applicable [diagnostic steps](https://github.com/vladmandic/human/wiki/Diag)

**Additional**

- For installation or startup issues include your `package.json`
- For usage issues, it is recommended to post your code as [gist](https://gist.github.com/)
- For general questions, create a [discussion topic](https://github.com/vladmandic/human/discussions)

**Comment by jpieters** [comment]
can you share the code?  I had some issues at first with Next.js 14, but got it working.  Ensure that you are explicitly using the "use client"; where appropriate.

**Comment by tembra** [comment]
@jpieters You can check the issue/example of the code in this repo: https://github.com/tembra/human-nextjs-14

There I tried many ways to load Human, each one in a specific intuitive named route, and they all fail.
- `http://localhost:3000/` default way
- `http://localhost:3000/no-ssr`
- `http://localhost:3000/lazy-loaded`
- `http://localhost:3000/load-on-button-click`

Notice that even on `/load-on-button-click` the error occurs during the build time. It seems that TypeScript/NextJS is executing the code inside the callback function that is assigned to a variable, but never called.

Even if NextJS is pre-rendering client components/pages on the server for the initial page load, it does not makes sense that this function is being executed.

Said that I think that `import` keyword is the problem, with TypeScript/NextJS trying to immediately find/import the module during the build time. Thinking this way I also tried to use (and also change/adapt) the webpack configuration from [next.config.js](https://github.com/vladmandic/human-next/blob/main/next.config.js) of the human-next demo.

If you can share how you got it working it will be awesome. I'm really out of ideas.

**Comment by tembra** [comment]
After digging deeper on webpack configuration I managed to get it working using this `next.config.mjs` configuration:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
    webpack: (config, { webpack, isServer }) => {
        if (isServer) {
            config.plugins.push(
                new webpack.IgnorePlugin({
                    checkResource(resource) {
                        if (resource === '@vladmandic/human') {
                            // ignore human package on server
                            return true
                        }

                        return false
                    },
                })
                // or a shorter (but not too explicit) version
                // new webpack.IgnorePlugin({ resourceRegExp: /^@vladmandic\/human$/ })
            )
        }

        return config
    }
}

export default nextConfig
```

With this configuration all my examples of the shared code work.

However I obviously got the error below during page compiling. So this is not the correct approach.

```bash
$ npm run dev

> human-nextjs-14@0.1.0 dev
> next dev

   ▲ Next.js 14.1.0
   - Local:        http://localhost:3000

 ✓ Ready in 6s
 ○ Compiling / ...
 ✓ Compiled / in 8.4s (477 modules)
 ⨯ Error: Cannot find module '@vladmandic/human'
    at webpackMissingModule (./app/page.tsx:7:50)
    at eval (./app/page.tsx:7:142)
    at (ssr)/./app/page.tsx (.../human-nextjs-14/.next/server/app/page.js:162:1)
    at __webpack_require__ (.../human-nextjs-14/.next/server/webpack-runtime.js:33:42)
    at JSON.parse (<anonymous>)
 ✓ Compiled in 243ms (230 modules)
```

I also still want to know WHY the module is being imported/resolved by webpack on server rendering if it is lazy loaded inside a callback.

Anyone has any thoughts/explanations or am I missing something in my code?

**Comment by vladmandic** [comment]
since this is not an issue with `human` itself, this is better suited for discussions.
also, i'm not a next.js user, so really cannot help much. i've experimented with next long time ago and while i liked the concept, i really did not like how its build/packaging process and weird webpack worked.

**Comment by alizaenazet** [comment]
# Solution for nextjs 14 app router client side browser as "use client" component.

> **Issue Description**
> 
> When importing Human on a NextJS 14 project, the framework tries to import the node bundle instead of ESM for browsers, despite adding `'use client'` on top of the page file.
> 
> Since node bundle requires a tfjs backend, the error below comes in:
> 
> ```
> ./node_modules/@vladmandic/human/dist/human.node.js:7:1678
> Module not found: Can't resolve '@tensorflow/tfjs-node'
> ```
> 

I encountered the same issue with the error `Module not found: Can't resolve '@tensorflow/tfjs-node'` when trying to use the model in the client-side browser with Next.js v14 App Router. My goal was to utilize the model entirely on the client side, with the `"use client"` directive.

**Cause :**
The issue arises because Next.js attempts to resolve modules during the build process before the page is accessible. However, since I wanted to load the model in the browser (client-side only), this led to the build-time error.

**Solution :**
To avoid this error, I decided with client-side approach using `componentDidMount` to dynamically load the `@vladmandic/human` library after the component is mounted in the browser. This ensures the code is only executed in the browser environment, bypassing the server-side build process.

### Example Implementation

```tsx
  override componentDidMount() { // make sure this will be ignored in build process
    if (typeof window !== 'undefined' && typeof document !== 'undefined') {
      // Assign img element after component is mounted
      this.img = document.getElementById(this.props.imgId) as (HTMLImageElement | undefined);
      
      // Load Human model dynamically
      import('@vladmandic/human').then((H: any) => {
        this.human = new H.default(config) as Human;
        console.log('human version:', this.human.version, '| tfjs version:', this.human.tf.version['tfjs-core']);
        console.log('platform:', this.human.env.platform, '| agent:', this.human.env.agent);
        console.log('loading models...');
        this.human.load().then(() => {
          this.setState({ ready: true });
          console.log('ready...');
        });
      });
    }
  }
```

### Key Points

1. **Dynamic Import in `componentDidMount`:** The dynamic import of `@vladmandic/human` is moved to `componentDidMount`, ensuring it only occurs after the component is mounted in the browser. This prevents errors during the server-side build process.
  
2. **Browser Environment Check:** The condition `if (typeof window !== 'undefined')` ensures that the code runs only in the browser, avoiding errors related to `document` or other browser-specific APIs during server-side rendering.

### Full Component Example

Below is the complete component that processes an `<img>` tag using the model upon clicking a button:

```tsx
'use client'
import { Component } from "react";
import type { Human, Config } from '@vladmandic/human';

const config: Partial<Config> = {
  debug: true,
  modelBasePath: '/models/models', // reference to my local models base path
  face: { enabled: true },
};

interface Props { imgId: string };
interface State { ready: boolean };

class RunHuman extends Component<Props, State> {
  human: Human | undefined = undefined;
  img: HTMLImageElement | undefined = undefined;
  setFaceResult: React.Dispatch<React.SetStateAction<Human[] | null>> | undefined;
  constructor(props: Props) {
    super(props);
    this.state = { ready: false };
  }

  override componentDidMount() { // make sure this will be ingored in build process
    if (typeof window !== 'undefined' && typeof document !== 'undefined') {
      // Assign img element after component is mounted
      this.img = document.getElementById(this.props.imgId) as (HTMLImageElement | undefined);
      
      // Load Human model dynamically
      import('@vladmandic/human').then((H: any) => {
        this.human = new H.default(config) as Human;
        console.log('human version:', this.human.version, '| tfjs version:', this.human.tf.version['tfjs-core']);
        console.log('platform:', this.human.env.platform, '| agent:', this.human.env.agent);
        console.log('loading models...');
        this.human.load().then(() => {
          this.setState({ ready: true });
          console.log('ready...');
        });
      });
    }
  }

  // Event handler for detection
  detect = async () => {
    if (!this.human || !this.img) return;

    // Run detection on the image
    await this.human.detect(this.img, { face: { detector: { maxDetected: 20 } } });

    // console.log detected faces and their embeddings
    const faces = this.human.result.face;
    console.log("total face detected: ",faces.length);
    faces.forEach((face, i) => {
        console.table({
            face: i+1, 
            gender: face.gender,
            score: face.score,
            embedding_length: face.embedding?.length
        })
    });

    console.log('Detection complete');
  };

  override render() {
    return (
      <div>
        <button className="p-2 mb-11 bg-black text-white border-red-400 border-2" onClick={this.detect} disabled={!this.state.ready}>
          Detect Faces
        </button>
      </div>
    );
  }
}

export default RunHuman;
```

This approach should help others facing the same issue when using TensorFlow.js models in client-side Next.js applications.
